services/galley/src/Galley/Effects/FederatorAccess.hs

```diff
@@ -0,0 +1,60 @@
+-- This file is part of the Wire Server implementation.
+--
+-- Copyright (C) 2021 Wire Swiss GmbH <opensource@wire.com>
+--
+-- This program is free software: you can redistribute it and/or modify it under
+-- the terms of the GNU Affero General Public License as published by the Free
+-- Software Foundation, either version 3 of the License, or (at your option) any
+-- later version.
+--
+-- This program is distributed in the hope that it will be useful, but WITHOUT
+-- ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+-- FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+-- details.
+--
+-- You should have received a copy of the GNU Affero General Public License along
+-- with this program. If not, see <https://www.gnu.org/licenses/>.
+
+module Galley.Effects.FederatorAccess
+  ( -- * Federator access effect
+    FederatorAccess (..),
+    runFederated,
+    runFederatedEither,
+    runFederatedConcurrently,
+    runFederatedConcurrently_,
+  )
+where
+
+import Data.Qualified
+import Galley.Intra.Federator.Types
+import Imports
+import Polysemy
+import Wire.API.Federation.Client
+import Wire.API.Federation.GRPC.Types
+
+data FederatorAccess m a where
+  RunFederated ::
+    forall (c :: Component) a m x.
+    Remote x ->
+    FederatedRPC c a ->
+    FederatorAccess m a
```

Comment:
Should this action eventually go, i.e., once we have error handling at the federator level? I suppose so, as a federated call can't always result in an `a`.

Reply:
I think it's fine to have this action, for code that doesn't really care about how the error is handled. That way the interpreter can decide how errors are ultimately thrown or reported and the user code doesn't need to know about errors when it uses the effect (unless it wants to handle them explicitly). But we'll have a better idea when we finish implementing the error effects.